bdk_electrum_example/src/electrum.rs

```diff
@@ -0,0 +1,197 @@
+            }
+        }
+
+        // Insert the new tip
+        let (tip_height, tip_hash) = self.get_tip()?;
+        if sparse_chain
+            .insert_checkpoint(BlockId {
+                height: tip_height,
+                hash: tip_hash,
+            })
+            .is_err()
+        {
+            // There has been an reorg since the last call to `block_header` in the loop above at the current tip.
+            return Err(ElectrumError::Reorg);
+        }
+
+        let mut keychain_index_update = BTreeMap::new();
+
+        // Fetch Keychain's last_active_index and all related txids.
+        // Add them into the KeyChainScan
+        for (keychain, mut scripts) in scripts {
+            let mut last_active_index = 0;
+            let mut unused_script_count = 0usize;
+
+            while let Some((index, script)) = scripts.next() {
+                let history = self
+                    .script_get_history(&script)?
+                    .iter()
+                    .map(|history_result| {
+                        if history_result.height > 0 && (history_result.height as u32) <= tip_height
+                        {
+                            return (
+                                history_result.tx_hash,
+                                TxHeight::Confirmed(history_result.height as u32),
+                            );
+                        } else {
+                            return (history_result.tx_hash, TxHeight::Unconfirmed);
+                        };
+                    })
+                    .collect::<Vec<(Txid, TxHeight)>>();
+
+                if history.is_empty() {
+                    unused_script_count += 1;
+                } else {
+                    last_active_index = index;
+                    unused_script_count = 0;
+
+                    for (txid, index) in history {
+                        if let Err(err) = sparse_chain.insert_tx(txid, index) {
+                            match err {
+                                InsertTxErr::TxTooHigh => {
+                                    unreachable!("We should not encounter this error as we ensured TxHeight <= tip_height");
+                                }
+                                InsertTxErr::TxMoved => {
+                                    /* This means there is a reorg, we will handle this situation below */
+                                }
+                            }
+                        }
+                    }
+                }
+
+                if unused_script_count >= stop_gap.unwrap_or(usize::MAX) {
+                    break;
+                }
+            }
+
+            keychain_index_update.insert(keychain, last_active_index);
+        }
+
+        // Check for Reorg during the above sync process
+        let tip_blockid = {
+            let (height, hash) = self.get_tip()?;
+            BlockId { height, hash }
+        };
+
+        if sparse_chain.latest_checkpoint().expect("must exist") != tip_blockid {
```

Comment:
This tip changing doesn't mean there has been a re-org. The block hash at the same height as `latest_checkpoint` changing means there has been a re-org.

Reply:
Yes makes sense.. Updated with hash matching of latest checkpoint that we have..